### PR TITLE
Feature/df 959 search profile

### PR DESCRIPTION
**Pull request by 99999989**

### CODE REVIEW

File: alv-portal-ui/src/app/dashboard/pav-dashboard/pav-dashboard.component.html [hunk 1]
```diff
@@ -10,6 +10,9 @@
 <alv-job-search-widget class="d-block mt-4">
 </alv-job-search-widget>
 
+<alv-job-search-profiles-widget class="d-block mt-4">
```
Comment (emina-str) [inline alv-portal-ui/src/app/dashboard/pav-dashboard/pav-dashboard.component.html]: I think the SP widget should be visible also on company and job-seeker dashboards.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      zipCode: location.postalCode,
+      geoPoint: location.coordinates,
+      regionCode: location.regionCode
+    };
+  }
+
+  private mapOccupations(occupations: OccupationTypeaheadItem[]): OccupationFilter[] {
+    return occupations.map(occupation => {
+      return {
+        labelId: occupation.payload.id // TODO: where to get the labelId from?
```
Comment (emina-str) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts]: I see this is working just fine, think you can remove the TODO.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      contractType: this.mapContractTypeToRequest(jobSearchFilter.contractType),
+      workloadPercentageMax: jobSearchFilter.workloadPercentageMax,
+      workloadPercentageMin: jobSearchFilter.workloadPercentageMin,
+      companyName: jobSearchFilter.company,
+      onlineSince: jobSearchFilter.onlineSince,
+      occupationFilters: this.mapOccupations(jobSearchFilter.occupations),
+      cantonFilters: this.mapCantons(jobSearchFilter.localities.filter(locality => locality.type === 'canton')),
+      localityFilters: this.mapLocalities(jobSearchFilter.localities.filter(locality => locality.type === 'locality')),
+      keywords: jobSearchFilter.keywords.map(keyword => keyword.payload).filter(k => !!k),
+      radiusSearchFilter: jobSearchFilter.localities.filter(locality => locality.type === 'locality').length === 1 ? {
```
Comment (emina-str) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts]: We're not mapping the coordinates, when it's one locality selected, to the geoPoints of the radiusSearchFilter.
It's not important for the result as we are mapping them in fromRequest and all is working fine, but we should map it to the radiusSearchFIlter so we can save it in the SEARCH_PROFILE table. Or another way would be to just remove this fields from the table in backend ... @pavax  what do you think?

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/update-search-profile-modal/update-search-profile-modal.component.html [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<alv-modal (primaryAction)="onUpdateExisting()"
```
Comment (emina-str) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/update-search-profile-modal/update-search-profile-modal.component.html]: I suppose primary and secondary actions order is correct like so, but in specification is other way around, so maybe to ask Ingo to replace the picture.

File: alv-portal-ui/src/app/shared/backend-services/job-ad-search-profiles/job-ad-search-profiles.repository.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      params,
+      observe: 'response'
+    }).pipe(
+      map((resp) => {
+        return {
+          totalCount: parseInt(resp.headers.get('X-Total-Count'), 10),
+          result: resp.body
+        };
+      }));
+    // return of({
```
Comment (emina-str) [inline alv-portal-ui/src/app/shared/backend-services/job-ad-search-profiles/job-ad-search-profiles.repository.ts]: remove?

File: alv-portal-ui/src/app/dashboard/pav-dashboard/pav-dashboard.component.html (hunk 1 above)
Comment (99999989) [reply alv-portal-ui/src/app/dashboard/pav-dashboard/pav-dashboard.component.html]: Good one. Added it to the other dashboards, too.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts (hunk 2 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts]: Removed

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/update-search-profile-modal/update-search-profile-modal.component.html (hunk 4 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/update-search-profile-modal/update-search-profile-modal.component.html]: I told Ingo, that the spec doesn't make sense regarding the order of the buttons. The title of the modal should relate to the primary button. 
The spec wants the primary button to look like the secondary button, too. I will discuss this matter with @pavax today.

File: alv-portal-ui/src/app/shared/backend-services/job-ad-search-profiles/job-ad-search-profiles.repository.ts (hunk 5 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/backend-services/job-ad-search-profiles/job-ad-search-profiles.repository.ts]: Removed

File: alv-portal-ui/src/app/shared/backend-services/job-advertisement/job-advertisement.types.ts [hunk 6]
```diff
@@ -337,3 +347,21 @@ export interface WebformPostAddress {
   city: string;
   countryIsoCode: string;
 }
+
+export interface JobAdSearchProfileResponse {
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/backend-services/job-advertisement/job-advertisement.types.ts]: let's define these types not in the job-advertisement.types.ts but rather in a job-profiles.types.ts

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts [hunk 7]
```diff
@@ -206,6 +234,67 @@ export class JobSearchComponent extends AbstractSubscriber implements OnInit, Af
...
+    modalRef.result
+      .then((searchProfile) => {
+        this.store.dispatch(new SearchProfileUpdatedAction({ searchProfile: searchProfile }));
+      })
+      .catch(() => {
+      });
+  }
+
+  deleteSearchProfile() {
+    this.store.pipe(
```
Comment (birom) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts]: There is no need to select the JobAdSearchProfileResponse again from the store. We can use the jobSearchProfile$ eg.:

```
 this.jobSearchProfile$.pipe(
      take(1)
    ).subscribe(searchProfile => {
...
```

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts [hunk 8]
```diff
@@ -206,6 +234,67 @@ export class JobSearchComponent extends AbstractSubscriber implements OnInit, Af
     return item.hashCode;
   }
 
+  saveSearchProfile() {
+    this.store.pipe(
```
Comment (birom) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts]: There is no need to select the JobAdSearchProfileResponse again from the store. We can use the jobSearchProfile$ eg.:

```
 this.jobSearchProfile$.pipe(
      take(1)
    ).subscribe(searchProfile => {
...
```

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts (hunk 7 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts]: Done

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts (hunk 8 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts]: done

File: alv-portal-ui/src/app/shared/backend-services/job-advertisement/job-advertisement.types.ts (hunk 6 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/backend-services/job-advertisement/job-advertisement.types.ts]: done

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/effects/job-ad-search.effects.ts [hunk 9]
```diff
@@ -232,6 +233,7 @@ export class JobAdSearchEffects {
               private jobAdFavouritesRepository: JobAdFavouritesRepository,
               private store: Store<JobAdSearchState>,
               private authenticationService: AuthenticationService,
+              private jobAdSearchProfilesRepository: JobAdSearchProfilesRepository,
```
Comment (birom) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/effects/job-ad-search.effects.ts]: We do not use this and the related import.

Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/effects/job-ad-search.effects.ts]: removed

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/update-search-profile-modal/update-search-profile-modal.component.html (hunk 4 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/update-search-profile-modal/update-search-profile-modal.component.html]: We agreed on keeping the current solution for timing reasons.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts (hunk 3 above)
Comment (emina-str) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/job-search-profile.service.ts]: will be removed from backend, we only need distance here.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/save-search-profile-modal/save-search-profile-modal.component.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        return this.jobAdSearchProfilesRepository.create(<CreateJobAdSearchProfile>{
+          name: this.form.get('name').value,
+          ownerUserId: currentUser.id,
+          searchFilter: this.jobSearchProfileService.mapToRequest(jobSearchFilter),
+        })
+          .pipe(
+            catchError(error => {
+              if (error.error.type) {
+
+                if (error.error.type === 'http://www.job-room.ch/job-ad-service/problem/search-profile/already-exists') {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/save-search-profile-modal/save-search-profile-modal.component.ts]: I would suggest an enum of such error codes.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.html [hunk 11]
```diff
@@ -65,14 +102,31 @@ <h2 class="sr-only">
...
+      <div class="w-100 mb-3 text-right text-md-left d-flex flex-column align-items-start">
         <a *ngIf="jobSearchMailToLink$ | async as jobSearchMailToLink"
-           class="btn btn-text btn-text-icon btn-truncate"
+           class="btn btn-text btn-text-icon btn-truncate mb-2"
            [href]="jobSearchMailToLink">
           <fa-icon class="mr-1"
                    [icon]="['fas', 'share']"></fa-icon>
           {{'portal.job-search.sendLink' | translate}}
         </a>
+        <button *ngIf="currentUser$ | async"
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.html]: double subscription to currentUser$ observable in this html file

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/reducers/job-ad-search.reducers.ts [hunk 12]
```diff
@@ -199,6 +201,17 @@ export function jobAdSearchReducer(state = initialState, action: Actions): JobAd
       break;
     }
 
+    case SEARCH_PROFILE_UPDATED: {
+      newState = {
+        ...state,
+        jobAdSearchProfile: action.payload.searchProfile,
+        details: {
+          ...state.details
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/reducers/job-ad-search.reducers.ts]: why is it needed? I thought when you do ...state, you already achieve this?

File: alv-portal-ui/src/app/widgets/job-search-profiles-widget/job-search-profiles-widget.component.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    );
+  }
+
+  onDeleteProfile(profile: JobAdSearchProfileResult) {
+    this.modalService.openConfirm(
+      getDeleteConfirmModalConfig(profile.name)
+    ).result
+      .then(result => {
+        this.jobAdSearchProfilesRepository.delete(profile.id)
+          .subscribe(() => {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/widgets/job-search-profiles-widget/job-search-profiles-widget.component.ts]: subscription inside a promise looks odd to me. I suggest using `fromPromise` function here and make the code less ladder-ish.

File: alv-portal-ui/translations.csv [hunk 14]
```diff
@@ -817,6 +817,27 @@ portal.job-detail.workExperience,Berufserfahrung,Work experience,Expérience pro
...
+portal.job-ad-search-profiles.save-modal.cancel-button.label,Abbrechen,Cancel,Annuler,Interrompere,
+portal.job-ad-search-profiles.notification.profile-saved,Ihr Suchprofil wurde gespeichert.,Your search has been saved.,Votre profil de recherche a été enregistré.,Il profilo di ricerca è stato salvato.,
+portal.job-ad-search-profiles.notification.profile-updated,Ihr Suchprofil wurde aktualisiert.,Your search has been updated.,Votre profil de recherche a été remplacé.,Il profilo di ricerca è stato sovrascritto.,
+portal.job-ad-search-profiles.notification.profile-deleted,Das Suchprofil wurde gelöscht.,Your saved search has been deleted.,Le profil de recherche a été supprimé.,Il profilo di ricerca è stato cancellato.,
+portal.job-ad-search-profiles.delete-confirmation-modal.question,"Möchten Sie das Suchprofil ""<strong>{{profileName}}</strong>"" wirklich löschen?","Are you sure you want to delete your search ""<strong>{{profileName}}</strong>""?","Souhaitez-vous vraiment supprimer le profil de recherche ""<strong>{{profileName}}</strong>""?","Volete davvero cancellare il profilo di ricerca ""<strong>{{profileName}}</strong>""?",
```
Comment (ganqqwerty) [inline alv-portal-ui/translations.csv]: is there a nicer way to have formatting in such texts without using html in the translation file?

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts [hunk 15]
```diff
@@ -206,6 +235,65 @@ export class JobSearchComponent extends AbstractSubscriber implements OnInit, Af
...
+    this.jobSearchProfile$.pipe(
+      take(1)
+    ).subscribe(searchProfile => {
+      if (searchProfile) {
+        this.modalService.openConfirm(
+          getDeleteConfirmModalConfig(searchProfile.name)
+        ).result
+          .then(result => {
+            this.jobAdSearchProfilesRepository.delete(searchProfile.id)
+              .subscribe(() => {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts]: promise/observable hell here

File: alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    });
+  }
+
+  onDeleteProfile(profile: JobAdSearchProfileResult) {
+    this.modalService.openConfirm(
+      getDeleteConfirmModalConfig(profile.name)
+    ).result
+      .then(result => {
+        this.jobAdSearchProfilesRepository.delete(profile.id)
+          .subscribe(() => {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts]: promise/observable hell here

File: alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+              private modalService: ModalService,
+              private notificationsService: NotificationsService) {
+  }
+
+  ngOnInit() {
+    this.onScroll();
+  }
+
+  onScroll() {
+    this.authenticationService.getCurrentUser().pipe(
```
Comment (pavax) [inline alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts]: there is no unsubscribe

File: alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts (hunk 16 above)
Comment (pavax) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts]: what would you suggest?

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/save-search-profile-modal/save-search-profile-modal.component.ts (hunk 10 above)
Comment (pavax) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/save-search-profile-modal/save-search-profile-modal.component.ts]: I agree let's add that enum the the profile.types.ts

File: alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts (hunk 16 above)
Comment (ganqqwerty) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search-profiles/job-ad-search-profiles/job-ad-search-profiles.component.ts]: either convert promise to observable with `fromPromise` function and chain it, or convert observable to promise with `toPromise` and chain it.

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.html (hunk 11 above)
Comment (pavax) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.html]: do not fix yet as we already have double subscriptions in this html file -> adress this problem later

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts (hunk 15 above)
Comment (pavax) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search/job-search.component.ts]: so what is your suggestion?

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/reducers/job-ad-search.reducers.ts (hunk 12 above)
Comment (pavax) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/reducers/job-ad-search.reducers.ts]: you are right I guess we can get rid of that

File: alv-portal-ui/src/app/widgets/job-search-profiles-widget/job-search-profiles-widget.component.ts (hunk 13 above)
Comment (pavax) [reply alv-portal-ui/src/app/widgets/job-search-profiles-widget/job-search-profiles-widget.component.ts]: I'd clean this up later

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/reducers/job-ad-search.reducers.ts (hunk 12 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/state-management/reducers/job-ad-search.reducers.ts]: removed

File: alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/save-search-profile-modal/save-search-profile-modal.component.ts (hunk 10 above)
Comment (99999989) [reply alv-portal-ui/src/app/job-advertisement/job-ad-search/job-search-profile/save-search-profile-modal/save-search-profile-modal.component.ts]: added

File: alv-portal-ui/src/app/widgets/job-search-profiles-widget/job-search-profiles-widget.component.ts (hunk 13 above)
Comment (99999989) [reply alv-portal-ui/src/app/widgets/job-search-profiles-widget/job-search-profiles-widget.component.ts]: This pattern is used everywhere in the application in conjunction with the confirm-modal. The ladder-ish look comes from the code formatting.

**Pull request merged**